californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -158,7 +163,8 @@ protected final Exchange matchNotifyResponse(final Response response, final Corr
 		final Exchange.KeyToken idByToken = Exchange.KeyToken.fromInboundMessage(response);
 		Exchange exchange = null;
 
-		final Observation obs = observationStore.get(response.getToken());
+		final EndpointIdentifier sourceEndpoint = response.getSourceEndpoint();
```

Comment:
This is set in [coapEndpoint](https://github.com/eclipse/californium/blob/endpoint_identifier/californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java#L748)

And the [`IdentifierFactory`](https://github.com/eclipse/californium/blob/endpoint_identifier/californium-core/src/main/java/org/eclipse/californium/core/identifier/IdentifierFactory.java) is responsible to define what should be used as Identifier.

(For Leshan I want to use Principal)